### PR TITLE
[Enhancement] reduce meaningless join transformation

### PR DESCRIPTION
**Pull request by packy92**
Fixes #issue

## Problem:
- Version 2.5, without outer join reorder, 3 inner joins, 4 left outer joins and join transformation takes about 1.5s
- Version 3.1, with outer join reorder enabled, 1 inner join, 4 left outer joins and join transformation takes about 114s

## root cause:
join transformation generated a large number of new groups due to split projection, but did not generate more valuable join orders.

## how to fix:
1. reduce meaningless join transformation by checking transformation mask
2. only rewriting the new on predicate like `A.col = B.col + C.col` to `A.col = add` when (B, C) is the bot join.
3. when computesize is smaller update group statistics
4. combine the SemiReorderRule with JoinLeftAsscomRule

## What type of PR is this:
- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Checklist:
- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.1
  - [x] 3.0
  - [ ] 2.5
  - [ ] 2.4

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 40 / 42 (95.24%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/JoinAssociativityRule.java|5|7|71.43%|[169, 175]|
|:large_blue_circle:|com/starrocks/sql/optimizer/Optimizer.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/task/DeriveStatsTask.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalJoinOperator.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/JoinLeftAsscomRule.java|9|9|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/RuleSet.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/join/JoinReorderProperty.java|10|10|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/statistics/StatisticsCalculator.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/RuleType.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/JoinAssociateBaseRule.java|3|3|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 51 / 62 (82.26%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/Memo.java|0|7|00.00%|[357, 358, 359, 360, 361, 363, 364]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/JoinAssociativityRule.java|5|9|55.56%|[172, 177, 182, 227]|
|:large_blue_circle:|com/starrocks/sql/optimizer/Optimizer.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/task/DeriveStatsTask.java|6|6|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalJoinOperator.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/JoinLeftAsscomRule.java|9|9|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/RuleSet.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/join/JoinReorderProperty.java|11|11|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/statistics/StatisticsCalculator.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/RuleType.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/JoinAssociateBaseRule.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/JoinCommutativityRule.java|6|6|100.00%|[]|

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/Optimizer.java [hunk 1]
```diff
@@ -526,11 +526,11 @@ void memoOptimize(ConnectContext connectContext, Memo memo, TaskContext rootTask
                 new ReorderJoinRule().transform(tree, context);
                 OptimizerTraceUtil.logOptExpression(connectContext, "after ReorderJoinRule:\n%s", tree);
 
-                context.getRuleSet().addJoinCommutativityWithOutInnerRule();
+                context.getRuleSet().addJoinCommutativityWithoutInnerRule();
             } else {
                 if (Utils.countJoinNodeSize(tree, JoinOperator.semiAntiJoinSet()) <
                         sessionVariable.getCboMaxReorderNodeUseExhaustive()) {
-                    context.getRuleSet().getTransformRules().add(new SemiReorderRule());
```
[the remark this answers is not in the document]
Comment (packy92) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/Optimizer.java:533]: Yes, the process the same pattern.

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/join/JoinReorderProperty.java [hunk 2]
```diff
@@ -67,6 +79,16 @@ private JoinReorderProperty() {
...
+        /*
+         *  bottom join operator within the row and top join operator with in the column
+         *  ▷* means null aware anti join
+         *   op    ⋈    ⟕   ⋉    ▷    ⋊    ◁    ⟖    ⟗   ×    ▷*
+         *  ----- ---- ---- ---- ---- ---- ---- ---- ---- ---- -----
+         *   ⋈     0    0    1    1    0    0    0    0    0    0
+         *   ×     0    0    1    1    0    0    0    0    0    0
+         */
+        INNER_LEFT_ASSCOM_PROPERTY[JoinOperator.INNER_JOIN.ordinal()] = new int[] {0, 0, 1, 1, 0, 0, 0, 0, 0, 0};
+        INNER_LEFT_ASSCOM_PROPERTY[JoinOperator.CROSS_JOIN.ordinal()] = new int[] {0, 0, 1, 1, 0, 0, 0, 0, 0, 0};
```
[the remark this answers is not in the document]
Comment (packy92) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/join/JoinReorderProperty.java:91]: The transformation you mentioned is generated by `JoinCommutativityRule`. The `JoinLeftAsscomRule` process pattern like (A join B) join C to (A join C) join B

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 53 / 64 (82.81%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/Memo.java|0|7|00.00%|[358, 359, 360, 361, 362, 364, 365]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/JoinAssociativityRule.java|6|10|60.00%|[170, 175, 180, 225]|
|:large_blue_circle:|com/starrocks/sql/optimizer/Optimizer.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/task/DeriveStatsTask.java|6|6|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalJoinOperator.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/JoinLeftAsscomRule.java|9|9|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/RuleSet.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/join/JoinReorderProperty.java|11|11|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/statistics/StatisticsCalculator.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/RuleType.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/JoinAssociateBaseRule.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/JoinCommutativityRule.java|6|6|100.00%|[]|

**Review by satanson — APPROVED** [review]

**Review by Seaven — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
@Mergifyio backport branch-3.1

**Comment by wanpengfei-git** [comment]
@Mergifyio backport branch-3.0

**Pull request merged**